### PR TITLE
Issue 11 MetricsSource shall send tuples either always or on changed value

### PR DESCRIPTION
**Pull request by m-kotowski**
Introduce the **emitMetricTuple** parameter for the MetricsSource operator as discussed in issue #11.

The allowed values for this parameter are:

* **periodic**: during each scan cycle emit a tuple for each monitored metric
* **onChangedValue** (default): during each scan cycle emit a tuple for each monitored metric if its value changed compared to its previous emitted value

### CODE REVIEW

File: com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MetricsSource.java [hunk 1]
```diff
@@ -155,6 +155,17 @@
...
+	private static final String DESC_PARAM_EMIT_METRIC_TUPLE =
+			"Specifies when to emit a tuple for a metric. Supported modes are:\\n"
+			+ "\\n"
+			+ "* **periodic**\\n"
+			+ "\\n"
+			+ "  For each monitored metric a tuple is emitted during each scan cycle.\\n"
+			+ "\\n"
+			+ "* **onChangedValue**\\n"
+			+ "\\n"
+			+ "  For each monitored metric a tuple is emitted during each scan cycle if the metric value changed.";
```
Comment (ddebrunner) [inline com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MetricsSource.java]: Description should indicate the default.

File: com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/DeltaMetricEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	}
+	
+	@Override
+	public boolean isSubmittable(Metric metric) {
+		return (_lastValue == null) || (Math.abs(_lastValue - metric.getValue()) >= _delta);
+	}
+
+	@Override
+	public void updateStatus(Metric metric) {
+		_lastValue = Long.valueOf(metric.getValue());
```
Comment (ddebrunner) [inline com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/DeltaMetricEvaluator.java]: Why is `Long.valueOf` needed, what type does `metric.getValue()` return?

File: com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/OperatorConfiguration.java [hunk 3]
```diff
@@ -161,4 +163,20 @@ public void set_tupleContainer(TupleContainer tupleContainer) {
...
+	public void set_emitMetricTuple(EmitMetricTupleMode mode) {
+		_emitMetricTuple = mode;
+	}
+
+	public IMetricEvaluator newDefaultMetricEvaluator() {
+		switch(_emitMetricTuple) {
+		case onChangedValue:
+			return new DeltaMetricEvaluator();
+		case periodic:
+			return new PeriodicMetricEvaluator();
```
Comment (ddebrunner) [inline com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/OperatorConfiguration.java]: Only a singleton `PeriodicMetricEvaluator` is required.

File: com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MetricsSource.java (hunk 1 above)
Comment (m-kotowski) [inline com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MetricsSource.java]: +1

File: com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/DeltaMetricEvaluator.java (hunk 2 above)
Comment (m-kotowski) [inline com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/DeltaMetricEvaluator.java]: +1
I was not aware that Java automatically creates a new object if a long is assigned to a Long variable.

File: com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/MetricOwningHandler.java [hunk 4]
```diff
@@ -56,15 +56,15 @@
 	/**
 	 * This map holds all metrics that are captured.
 	 */
-	private Map<String /* metric name */, Metric> _capturedMetrics = new HashMap<String, Metric>();
+	private Map<String /* metric name */, IMetricEvaluator> _capturedMetrics = new HashMap<String, IMetricEvaluator>();
```
Comment (m-kotowski) [inline com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/MetricOwningHandler.java]: +1
Nice feature.

File: com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/OperatorConfiguration.java (hunk 3 above)
Comment (m-kotowski) [inline com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/OperatorConfiguration.java]: +1

**Pull request merged**